redash/query_runner/big_query.py

```diff
@@ -246,22 +250,106 @@ def _get_query_result(self, jobs, query):
 
     def _get_columns_schema(self, table_data):
         columns = []
+        metadata = []
         for column in table_data.get('schema', {}).get('fields', []):
-            columns.extend(self._get_columns_schema_column(column))
+            column, metadatum = self._get_columns_schema_column(column)
+            metadata.extend(metadatum)
+            columns.extend(column)
 
         project_id = self._get_project_id()
         table_name = table_data['id'].replace("%s:" % project_id, "")
-        return {'name': table_name, 'columns': columns}
+        return {'name': table_name, 'columns': columns, 'metadata': metadata}
 
     def _get_columns_schema_column(self, column):
         columns = []
+        metadata = []
+
         if column['type'] == 'RECORD':
             for field in column['fields']:
-                columns.append(u"{}.{}".format(column['name'], field['name']))
+                col_name = u"{}.{}".format(column['name'], field['name'])
+                columns.append(col_name)
+                metadata.append({'name': col_name, 'type': field['type']})
         else:
             columns.append(column['name'])
+            metadata.append({'name': column['name'], 'type': column['type']})
+
+        return columns, metadata
```

Comment:
I'm not a huge fan when a method is extended by returning tuples of values that are related but not quite the same since it makes type checking of the return value harder and encourages wrong data unpacking, or eve more extended return types in the future.

And since it seems this method isn't used elsewhere, can we just make this a `_get_column_metadata` method that then returns the list of metadata for the given column including the name that `_get_columns_schema` needs?

In `_get_columns_schema` the `columns` value could be extended with `columns.extend(map(operator.itemgetter('name'), metadatum))` then. What do you think?

Reply:
I like this idea, thanks!